entity-framework/core/testing/choosing-a-testing-strategy.md

```diff
@@ -0,0 +1,112 @@
+
+1. Most databases can nowadays be easily installed on the developer's machine. Container-based technologies such as Docker can make this very easy, and technologies such as [Github Workspaces](https://docs.github.com/en/codespaces/overview) and [Dev Container](https://code.visualstudio.com/docs/remote/create-dev-container) set up your entire development environment for you (including the database). When using SQL Server, it's also possible to test against [LocalDB](/sql/database-engine/configure-windows/sql-server-express-localdb) on Windows, or easily set up a Docker image on Linux.
+2. Testing against a local database - with a reasonable test dataset - is usually extremely fast: communication is completely local, and test data is typically buffered in memory on the database side. EF Core itself contains over 30,000 tests against SQL Server alone; these complete reliably in a few minutes, execute in CI on every single commit, and are very frequently executed by developers locally. Some developers turn to an in-memory database (a "fake") in the belief that this is needed for speed - this is almost never actually the case.
+3. Isolation is indeed a hurdle when running tests against a real database, as tests may modify data and interfere with one another. However, there are various techniques to provide isolation in database testing scenarios; we concentrate on these in the [integration testing docs](xref:core/testing/integration-testing)).
+
+The above is not meant to disparage unit testing or to argue against writing them. For one thing, unit tests are necessary for some scenarios which cannot be tested with integration tests, such as simulating database failure. However, in our experience, users frequently shy away from integration testing for the above reasons, believing it's slow, hard or unreliable, when that isn't necessarily the case. [The dedicated page on integration testing](xref:core/testing/integration-testing) aims to address this, providing guidelines and samples for writing fast, isolated database integration tests.
+
+## Unit testing
+
+Unlike integration tests, unit tests never involve the actual database system used in production, but rather swap it with a [test double](https://en.wikipedia.org/wiki/Test_double). Below are some common methods for unit testing EF Core applications with test doubles:
+
+1. Use SQLite (in-memory mode) as a database fake, replacing your production database system.
+2. Use the EF Core InMemory provider as a database fake, replacing your production database system.
+3. Mock or stub out `DbContext` and `DbSet`.
+4. Introduce a repository layer between EF Core and your application code, and mock or stub that layer.
+
+Below, we'll explore what each method means, and compare it with the others. We recommend reading through the different methods to gain a full understanding, but if you're looking for a quick suggestion, we highly recommend the 4th method: introducing a repository layer.
```

Comment:
I still don't think we should say we highly recommend a repository layer.

Reply:
Agreed. Especially with how much the ASP.NET team is trying to simplify application boilerplate and bootstrap.

It would for sure be jarring to someone getting started with .NET server side development.  Consider from their perspective:  Going from a tutorial on ASP.NET telling them how they've lowered the ceremony and then arrive at Entity Framework docs which tell them to strap in for a lot of repository pattern boilerplate.

Most people just want to write LINQ.  The state of the art here is LINQ in mediator listeners/handlers.

Reply:
This statement is made within the context of the "Unit testing" section, but I agree and will tweak to make this more explicit and nuanced.

@atrauzzi I don't disagree, but that's the reality of the situation. If you (a) insist on doing unit tests (no database involved), and (b) do anything beyond a very restricted subset of queries, then the only way to do this reliably and comprehensively is via a repository layer. If that's unacceptable because it's too complicated/ceremonious, you can test against your real database - but within the above two constraints there's not much else we can say.

Reply:
Here's what I ended up with for now:

> Below, we'll explore what each method means, and compare it with the others; We recommend reading through the different methods to gain a full understanding of each one. If you've decided that unit tests are required (as opposed to integration tests), than a repository layer is the only approach allowing the comprehensive and reliable stubbing/mocking of the data layer. However, that approach has a significant cost in terms of implementation and maintenance.

Reply:
Makes sense, and I've come to agree. Using the real database is essential without a repository layer. 👍